centinel/primitives/traceroute.py

```diff
@@ -0,0 +1,131 @@
+#
+# Abbas Razaghpanah (arazaghpanah@cs.stonybrook.edu)
+# February 2015, Stony Brook University
+#
+# tracerout.py: interface to system traceroute
+# for this to work, traceroute has to be installed
+# and accessible.
+
+
+import time
+
+
+from centinel import command
+
+
+def traceroute(url, method="icmp", cmd_arguments=[]):
+    """This function uses centinel.command to issue
+    a traceroute command, wait for it to finish execution and
+    parse the results out to a dictionary.
+
+    Params:
+    url- the URL to be queried
+
+    method- the packet type used for traceroute, ICMP by default
+
+    cmd_arguments- the list of arguments that need to be passed
+    to traceroute.
+
+    """
+
+    # the method specified by the function parameter here will
+    # over-ride the ones given in cmd_arguments because
+    # traceroute will use the last one in the argument list.
+
+    if method == "tcp":
+        cmd_arguments.append('-T')
+    elif method == "udp":
+        cmd_arguments.append('-U')
+    elif method == "icmp":
+        cmd_arguments.append('-I')
+
+    cmd = ['traceroute'] + cmd_arguments + [url]
+    caller = command.Command(cmd, _traceroute_callback)
+    caller.start()
+    if not caller.started:
+        message = ""
+        if caller.exception is not None:
+            if "No such file or directory" in caller.exception:
+                message = ": traceroute not found or not installed"
+            else:
+                message = (", traceroute thread threw an exception: " +
+                       str(caller.exception))
+        if "enough privileges" in caller.notifications:
```

Comment:
It never happens. The exception is thrown when the process fails to start, whereas the errors in notifications are passed from the already-started process. So if an exception is thrown, you will not see any other error because traceroute process has not started.
When lacking privileges, traceroute closes. It does not try to resolve the domain name after that.